### PR TITLE
IndexJoin: Track number of chunks that actually used index

### LINKED ISSUE #759: IndexJoin: Track usage for optimization and tests
Right now, the `IndexJoin` could fall back to a nested loop join if no index exists. We should track the number of chunks that were joined using an index so that we can (a) show a warning if the optimizer screwed up, and (b) let the tests make sure that the index was actually used.

### PR DESCRIPTION
**Pull request by mrks**
Also gives an example on how operators can implement their own performance data. fixes #759

### CODE REVIEW

File: src/lib/operators/join_index.cpp [hunk 1]
```diff
@@ -182,6 +187,12 @@ void JoinIndex::_perform_join() {
   _write_output_columns(output_columns, _right_in_table, _pos_list_right);
 
   _output_table->append_chunk(output_columns);
+
+  if (performance_data.chunks_scanned_with_index > performance_data.chunks_scanned_without_index) {
```
Comment (Bensk1) [inline src/lib/operators/join_index.cpp]: Is this correct? I somehow expected the comparison to be the other way around. Also, should it display the total number of chunks after the `of` then it should be `performance_data.chunks_scanned_without_index + performance_data.chunks_scanned_with_index`?

File: src/lib/operators/join_index.hpp [hunk 2]
```diff
@@ -27,6 +27,13 @@ class JoinIndex : public AbstractJoinOperator {
 
   const std::string name() const override;
 
+  struct PerformanceData : public OperatorPerformanceData {
```
Comment (Bensk1) [inline src/lib/operators/join_index.hpp]: Could we call that `Join`- or `JoinIndex`PerformanceData? `PerformanceData` inheriting from `OperatorPerformanceData` seems confusing.

File: src/lib/operators/operator_performance_data.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#include "types.hpp"
+
+namespace opossum {
+
+// For an example on how this can be extended on a per-operator basis, see JoinIndex
+
+struct OperatorPerformanceData : public Noncopyable {
+  virtual ~OperatorPerformanceData();
+
+  std::chrono::microseconds walltime{0};
```
Comment (Bensk1) [inline src/lib/operators/operator_performance_data.hpp]: Does it make sense to store the wall time in microseconds when we print nanoseconds later?

File: src/lib/operators/join_index.hpp (hunk 2 above)
Comment (mrks) [reply src/lib/operators/join_index.hpp]: The full name is `JoinIndex::PerformanceData`, you can only use `PerformanceData` within `JoinIndex`. Did you take that into account?

File: src/lib/operators/operator_performance_data.hpp (hunk 3 above)
Comment (mrks) [reply src/lib/operators/operator_performance_data.hpp]: That's what we get from the `Timer` class. We don't print ns but convert it into the most readable format. ns is just for the interface of `format_duration`.

File: src/lib/operators/join_index.hpp (hunk 2 above)
Comment (Bensk1) [reply src/lib/operators/join_index.hpp]: No, never mind then.

**Pull request merged**